assets/src/models/ladderVehicle.ts

```diff
@@ -0,0 +1,239 @@
+ *
+ * Steps:
+ *  1) Calculate VehicleOnLadder --> the properties that don't depend on anything else
+ *  2) Calculate the lane property
+ *  3) Calculate the x property
+ */
+export const ladderVehiclesFromVehicles = (
+  vehicles: Vehicle[],
+  ladderDirection: LadderDirection,
+  yFunc: YFunc
+): {
+  ladderVehicles: LadderVehicle[]
+  widthOfLanes: number
+} => {
+  const vehiclesOnLadder: VehicleOnLadder[] = vehicles.map(vehicle =>
+    vehicleOnLadder(vehicle, ladderDirection, yFunc)
+  )
+
+  const vehiclesInLane: VehicleInLane[] = putIntoLanes(vehiclesOnLadder)
+
+  const maxOccupiedLane: number = numOccupiedLanes(vehiclesInLane)
+  const ladderVehicleHorizontalOffset: number = horizontalOffsetForLanes(
+    maxOccupiedLane
+  )
+  const xInLane = xInLaneWithOffset(ladderVehicleHorizontalOffset)
+  // const widthOfLanes = maxOccupiedLane * ladderVehicleHorizontalOffset
+  const widthOfLanes =
+    widthOfVehicleGroup + (maxOccupiedLane - 1) * ladderVehicleHorizontalOffset
+
+  const ladderVehicles: LadderVehicle[] = vehiclesInLane.map(vehicleInLane =>
+    addX(vehicleInLane, xInLane)
+  )
+
+  return {
+    ladderVehicles,
+    widthOfLanes,
+  }
+}
+
+export const vehicleDirectionOnLadder = (
+  vehicle: Vehicle,
+  ladderDirection: LadderDirection
+): VehicleDirection =>
+  (vehicle.directionId === 1) ===
+  (ladderDirection === LadderDirection.ZeroToOne)
+    ? VehicleDirection.Down
+    : VehicleDirection.Up
+
+const vehicleOnLadder = (
+  vehicle: Vehicle,
+  ladderDirection: LadderDirection,
+  yFunc: YFunc
+): VehicleOnLadder => {
+  const vehicleDirection: VehicleDirection = vehicleDirectionOnLadder(
+    vehicle,
+    ladderDirection
+  )
+
+  const y = yFunc(vehicle.timepointStatus, vehicleDirection) || -10
+
+  return {
+    // tslint:disable-next-line:object-literal-sort-keys
+    vehicle,
+    vehicleDirection,
+    y,
+  }
+}
+
+export const putIntoLanes = (
+  vehiclesOnLadder: VehicleOnLadder[]
+): VehicleInLane[] =>
+  vehiclesOnLadder
+    .slice()
+    .sort(byDirectionAndY)
+    .reduce(
+      (
+        previousLadderVehiclePositions: VehicleInLane[],
+        ladderVehicle: VehicleOnLadder
+      ): VehicleInLane[] => {
+        const overlappers: InLane[] = overlappingPreviousVehicles(
+          previousLadderVehiclePositions,
+          ladderVehicle
+        )
+        const occupiedLanes: number[] = overlappers.map(
+          overlapper => overlapper.lane
+        )
+        const lane = firstOpenLane(occupiedLanes)
+
+        return [
+          {
+            ...ladderVehicle,
+            // tslint:disable-next-line:object-literal-sort-keys
+            lane,
+          },
+          ...previousLadderVehiclePositions,
+        ]
+      },
+      []
+    )
+
+const numOccupiedLanes = (vehicles: InLane[]): number => {
+  const allLanes: number[] = vehicles.map(({ lane }) => lane || 0)
+  const furthestLane = allLanes.length ? Math.max(...allLanes) : 0
+  return furthestLane + 1
```

Comment:
`numOccupiedLanes` and `maxOccupiedLanes` get conflated. Since one of the uses subtracts 1 later, I think this function should get renamed `maxOccupiedLane` and the `+ 1` here removed.

The `- 1` when calculated `widthOfLanes` can be removed then. (The `horizontalOffsetForLanes` also uses `maxOccupiedLane`, and might need to tweak its calculations.)